server/appservices/subscriptions.go

```diff
@@ -19,69 +20,193 @@ type PermissionChecker interface {
+	return err
 }
 
-func (a *AppServices) Subscribe(_ *incoming.Request, sub apps.Subscription) error {
-	if err := sub.Validate(); err != nil {
-		return utils.NewInvalidError("invalid subscription")
+func (a *AppServices) unsubscribe(r *incoming.Request, ownerUserID string, e apps.Event) ([]store.Subscription, error) {
+	all, err := a.store.Subscription.Get(e)
+	if err != nil {
+		return nil, err
 	}
 
-	if err := CheckSubscriptionPermission(&a.conf.MattermostAPI().User, sub, "", ""); err != nil {
-		return err
+	for i, s := range all {
+		if s.AppID != r.SourceAppID() || s.OwnerUserID != ownerUserID {
+			continue
+		}
+
+		allowed := false
+		if s.OwnerUserID == r.ActingUserID() {
+			allowed = true
+		} else if err = r.RequireSysadminOrPlugin(); err == nil {
+			allowed = true
+		}
+		if !allowed {
+			return nil, utils.NewForbiddenError("must be the owner of the subscription, or system administrator to unsubscribe")
+		}
+
+		modified := all[:i]
+		if i < len(all) {
+			modified = append(modified, all[i+1:]...)
+		}
+		err = a.store.Subscription.Save(e, modified)
+		if err != nil {
+			return nil, err
+		}
+		return modified, nil
 	}
 
-	return a.store.Subscription.Save(sub)
+	return all, utils.ErrNotFound
 }
 
-func (a *AppServices) GetSubscriptions(_ *incoming.Request, appID apps.AppID, userID string) ([]apps.Subscription, error) {
-	return a.store.Subscription.ListByUserID(appID, userID)
-}
+func (a *AppServices) hasPermissionToSubscribe(r *incoming.Request, sub apps.Subscription) func() error {
+	return func() error {
+		mm := r.Config().MattermostAPI()
+		userID := r.ActingUserID()
 
-func (a *AppServices) Unsubscribe(_ *incoming.Request, sub apps.Subscription) error {
-	if err := sub.Validate(); err != nil {
-		return utils.NewInvalidError("invalid subscription")
-	}
+		switch sub.Subject {
+		case apps.SubjectUserCreated:
+			if !mm.User.HasPermissionTo(userID, model.PermissionViewMembers) {
+				return errors.New("no permission to read user")
+			}
 
-	return a.store.Subscription.Delete(sub)
+		case apps.SubjectUserJoinedChannel, apps.SubjectUserLeftChannel /*, apps.SubjectPostCreated, apps.SubjectBotMentioned */ :
+			if !mm.User.HasPermissionToChannel(userID, sub.ChannelID, model.PermissionReadChannel) {
+				return errors.New("no permission to read channel")
+			}
+
+		case apps.SubjectUserJoinedTeam, apps.SubjectUserLeftTeam:
+			if !mm.User.HasPermissionToTeam(userID, sub.TeamID, model.PermissionViewTeam) {
+				return errors.New("no permission to view team")
+			}
+
+		case apps.SubjectBotJoinedChannel,
+			apps.SubjectBotLeftChannel,
+			apps.SubjectBotJoinedTeam,
+			apps.SubjectBotLeftTeam:
+			// When the bot has joined an entity, it will have the permission to
+			// read it.
+			return nil
+
+		case apps.SubjectChannelCreated:
+			if !mm.User.HasPermissionToTeam(userID, sub.TeamID, model.PermissionListTeamChannels) {
+				return errors.New("no permission to list channels")
+			}
```

Comment:
Wondering if this fires off for private channels

Reply:
We could add a test, but if the subscription was created by (and is executed as) a user that has access to the new channel (the author, sysadmin, ?..) it should work. Otherwise, it should fail for lack of channel membership.